[AutoDiff] Derive `Differentiable.zeroTangentVectorInitializer`.

dan-zheng opened the pull request:
`Differentiable` conformance derivation now supports
`Differentiable.zeroTangentVectorInitializer`.

`zeroTangentVectorInitializer` is a closure that produces a zero tangent vector,
capturing minimal necessary information from `self`.

It is an instance property, unlike the static property `AdditiveArithmetic.zero`,
and should be used by the differentiation transform for correctness.

Remove `zeroTangentVectorInitializer` dummy default implementation.
Primitive `Differentiable`-conforming types now must implement it.

Update stdlib `Differentiable` conformances and tests.

Resolves TF-1007.
Progress towards TF-1008: differentiation correctness for projection operations.

---

Examples:

```swift
import _Differentiation

struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Compiler synthesizes:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     return TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}

let s = Struct(x: [1, 2, 3], y: [[4, 5, 6], [], [2]])
// `Differentiable.zeroTangentVector` default implementation calls
// `zeroTangentVectorInitializer`.
print(s.zeroTangentVector)

// Before (via dummy `zeroTangentVectorInitializer` default implementation):
// TangentVector(x: [], y: [])

// After (via derived conformances):
// TangentVector(x: [0.0, 0.0, 0.0], y: [[0.0, 0.0, 0.0], [], [0.0]])
```

```swift
struct CustomTangentVector<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  typealias TangentVector = T.TangentVector
  mutating func move(along direction: TangentVector) {}

  // Memberwise synthesis is not possible for custom `TangentVector` type.
  // Fallback synthesis is done instead:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { TangentVector.zero }
  // }
}
```

dan-zheng commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -1449,6 +1449,16 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
     /// use or capture of "self." for qualification of member references.
     static bool isClosureRequiringSelfQualification(
                   const AbstractClosureExpr *CE) {
+      // If the closure is in an implicit declaration, then it doesn't need
+      // qualification. Diagnostics would show an invalid source location.
+      //
+      // Currently used for `Differentiable.zeroTangentVectorInitializer`
+      // derived conformances.
+      if (CE->isImplicit()) {
+        auto *decl = CE->getInnermostDeclarationDeclContext();
+        if (decl && decl->isImplicit())
+          return false;
+      }
```
Note: this special case is necessary to avoid diagnosing implicit `self` `DeclRefExpr` qualification in synthesized `zeroTangentVectorInitializer` declarations.

marcrasi replied:
Is it possible to synthesize `self.xxx` instead of adding this special case here?

dan-zheng replied:
> Is it possible to synthesize `self.xxx` instead of adding this special case here?

`self.xxx` was actually synthesized as an implicit `DeclRefExpr`. However, there's no distinction between an implicit `self` `DeclRefExpr` generated via derived conformances or via other parts of type-checking, which made this special case necessary.

dan-zheng replied:
I was able to revert this special case!

The previously synthesized ASTs were actually wrong. Synthesized `Differentiable.zeroTangentVectorInitializer` closures should never capture `self` (i.e. contain implicit `self` `DeclRefExpr`s).

```swift
import _Differentiation
struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Previously synthesized, incorrectly captures `self`:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   {
  //     // Accesses to `self.x` and `self.y` cause `self` to be captured.
  //     TangentVector(x: x.zeroTangentVectorInitializer(),
  //                   y: y.zeroTangentVectorInitializer())
  //   }
  // }

  // Now synthesized, correct:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}
```

dan-zheng commented on lib/Sema/DebuggerTestingTransform.cpp:
```diff
@@ -33,35 +34,6 @@ using namespace swift;
-///
-/// The parser typically takes care of assigning unique discriminators to
-/// closures, but the parser is unavailable to this transform.
-class DiscriminatorFinder : public ASTWalker {
```
Note: I moved `DiscriminatorFinder` to a shared location `lib/Sema/CodeSynthesis.h`.

dan-zheng commented:
@CodaFi: I'll add you as reviewer in case you'd like to look at any of these derived conformances changes. `lib/Sema/DerivedConformanceDifferentiable.cpp` has some room for cleanup that I'd like to investigate later.

dan-zheng commented:
@swift-ci Please test

rxwei commented on test/AutoDiff/Sema/DerivedConformances/class_differentiable.swift:
```diff
@@ -109,28 +109,13 @@ final class VectorSpacesEqualSelf: Differentiable & DummyAdditiveArithmetic {
   var w: Float
   var b: Float
   typealias TangentVector = VectorSpacesEqualSelf
+  var zeroTangentVectorInitializer: () -> TangentVector { fatalError() }
```
Shouldn't these use the default implementation (`.zero`)?  Is it because derivation won't be triggered if you provide a default implementation?

dan-zheng replied:
> Is it because derivation won't be triggered if you provide a default implementation?

Yes, we have to remove the dummy default implementation of `Differentiable.zeroTangentVectorInitializer` using `TangentVector.zero` because its existence prevents derived conformances from triggering.

This requires "primitive" `Differentiable`-conforming types to implement `zeroTangentVectorInitializer`.

rxwei commented on include/swift/AST/KnownIdentifiers.def:
```diff
@@ -223,6 +223,8 @@ IDENTIFIER(move)
 IDENTIFIER(pullback)
 IDENTIFIER(TangentVector)
 IDENTIFIER(zero)
+IDENTIFIER(zeroTangentVector)
```
Why is this needed?

dan-zheng replied:
Thanks, this isn't needed. Removed.

rxwei reviewed (commented):
Can you derive a default implementation for `zeroTangentVectorInitializer` even when the user has a custom tangent vector? If you do that then this PR won't be source-breaking.

dan-zheng commented:
> Can you derive a default implementation for `zeroTangentVectorInitializer` even when the user has a custom tangent vector? If you do that then this PR won't be source-breaking.

I asked offline for clarification. What @rxwei has in mind is: deriving `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`.

@marcrasi: do you agree whether this makes sense and is desirable? I think you have experience with custom `TangentVector` types from [borglab/SwiftFusion](https://github.com/borglab/SwiftFusion).

This behavior should be possible but requires more work. I'd like to reach consensus before exploring.

marcrasi commented:
I have some thoughts about default synthesis of zeroTangentVectorInitializer.

**The ultimate ideal behavior**

* We clarify the semantics of `AdditiveArithmetic` to require that `static var zero` always "makes sense". (Depending on how you interpret the documentation, this is already how the semantics are, and it is wrong for `Tensor` to conform to `AdditiveArithmetic`).
* Therefore we remove the conformances of `Array`, `Tensor`, etc to `AdditiveArithmetic`.
* We introduce a new `protocol X` that captures the concept of an additive arithmetic without zero. `AdditiveArithmetic` refines `X`.
* We relax the `Differentiable` requirement that `TangentVector: AdditiveArithmetic`. Make it `TangentVector: X` instead.
* When `Differentiable` synthesis sees a custom `TangentVector` that conforms to `AdditiveArithmetic`, it synthesizes a `zeroTangentVectorInitializer`. This is completely correct and safe because of the semantics of `AdditiveArithmetic`.
* When `Differentiable` synthesis sees a custom `TangentVector` that does not conform to `AdditiveArithmetic`, it is unable to synthesize `zeroTangentVectorInitializer`.

**What to do today**

Yes, derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`. This is consistent with the ultimate ideal behavior, and therefore a step in the right direction.

marcrasi commented:
**From a usability point of view**

It's already pretty difficult to define a custom `TangentVector`. If we don't automatically synthesize `zeroTangentVectorInitialzer` for custom `TangentVector`s, then it becomes even more difficult. This is a step in the wrong direction for usability.

Therefore the usability point of view also says that, it's good to derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`.

marcrasi commented on stdlib/public/Differentiation/ArrayDifferentiation.swift:
```diff
@@ -168,9 +172,7 @@ extension Array: Differentiable where Element: Differentiable {
-    { [count = self.count] in
-      TangentVector(.init(repeating: .zero, count: count))
-    }
+    { TangentVector(map { $0.zeroTangentVector }) }
```
This is interesting, good catch.

However, this is capturing `self`. It should instead capture all the elements' `zeroTangentVectorInitializer`s.

This sounds potentially terrible for performance for a few reasons:

* If the closure captures anything, then it probably does one heap allocation per closure. (Maybe the compiler is clever enough to realize that all the closure captures are the same and only do one heap allocation for all of them? That would be very magical.
* In the case where `TangentVector(.init(repeating: .zero, count: count))` is correct, we would really like to use `count` so that we can avoid iterating over the array.

I can think of one solution that involves defining a few different differentiable views of array that are specialized to work efficiently in different cases. This solution can be implemented in client code, so we could leave the stdlib as-is for now and eventually move the solution into the stdlib if we discover that it really is necessary.

I wonder if there are any other solutions that are easier for users to use?

rxwei replied:
Yeah, calling `.zeroTangentVector` defeats the whole purpose of `zeroTangentVectorInitializer`. How about:
```swift
{ [zeroInits = map(\.zeroTangentVectorInitializer)] in
    TangentVector(zeroInits.map { $0() })
}
```

dan-zheng replied:
Thanks for the catch!

I applied @rxwei's suggestion: `Array.zeroTangentVectorInitializer` captures only elements' `zeroTangentVectorInitializer` closures.

marcrasi commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -281,6 +299,139 @@ deriveBodyDifferentiable_move(AbstractFunctionDecl *funcDecl, void *) {
+  return std::pair<BraceStmt *, bool>(braceStmt, false);
+}
+
+/// Synthesize body for `move(along:)`.
```
This comment is wrong?

dan-zheng replied:
Fixed, thanks!

marcrasi commented on lib/Sema/DerivedConformanceDifferentiable.cpp:
```diff
@@ -281,6 +299,139 @@ deriveBodyDifferentiable_move(AbstractFunctionDecl *funcDecl, void *) {
 
+/// Synthesize body for a `Differentiable` method requirement.
+static std::pair<BraceStmt *, bool>
+deriveBodyDifferentiable_zeroTangentVectorInitializer_getter(
```
It would be useful to have comments inline with pseudocode corresponding to the code that is being generated.

There is some pseudocode but it's confusingly unrelated to the actual code that is being generated. Like the "`x.move(along: direction.x)`". I don't think it's synthesizing any `move(along:)` calls here.

dan-zheng replied:
Thanks for the catch! I copied the code for synthesized `move(along:)` and forgot to update comments.

Comments are fixed now.

rxwei commented:
The root problem is that types like `Tensor` have dynamic ranks, so the tangent vector type of a `Tensor` is in fact value-dependent.  I wish Swift supports things like path-dependent types one day.

```swift
@derivative(of: foo)
func _(x: T) -> (value: U, differential: @differentiable(linear) (x.T) -> value.TangentVector
```

However, `zeroTangentVectorInitializer()` will be a required workaround in the foreseeable future...

dan-zheng committed:
[AutoDiff] Derive `Differentiable.zeroTangentVectorInitializer`.

`Differentiable` conformance derivation now supports
`Differentiable.zeroTangentVectorInitializer`.

There are two potential cases:
1. Memberwise derivation: done when `TangentVector` can be initialized memberwise.
2. `{ TangentVector.zero }` derivation: done as a fallback.

`zeroTangentVectorInitializer` is a closure that produces a zero tangent vector,
capturing minimal necessary information from `self`.

It is an instance property, unlike the static property `AdditiveArithmetic.zero`,
and should be used by the differentiation transform for correctness.

Remove `Differentiable.zeroTangentVectorInitializer` dummy default implementation.

Update stdlib `Differentiable` conformances and tests.
Clean up DerivedConformanceDifferentiable.cpp cruft.

Resolves TF-1007.
Progress towards TF-1008: differentiation correctness for projection operations.

dan-zheng commented:
Thanks for the feedback above!

> Yes, derive `zeroTangentVectorInitializer` as `{ TangentVector.zero }` for types defining a custom `TangentVector`. This is consistent with the ultimate ideal behavior, and therefore a step in the right direction.

I implemented this behavior. There are now two cases for `Differentiable.zeroTangentVectorInitializer` derivation:
1. Memberwise derivation, when `TangentVector` type can be initialized memberwise.
2. Fallback `{ TangentVector.zero }` derivation.

The result is improved usability: almost no types (including "primitive" `Differentiable`-conforming ones) are _required_ to define `Differentiable.zeroTangentVectorInitializer`.

---

`Differentiable.zeroTangentVectorInitializer` derivation is now also fixed so that synthesized closures never capture `self`.

<details>

```swift
import _Differentiation
struct Struct<T: Differentiable, U: Differentiable>: Differentiable {
  var x: T
  var y: U

  // Previously synthesized, incorrectly captures `self`:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   {
  //     // Accesses to `self.x` and `self.y` cause `self` to be captured.
  //     TangentVector(x: x.zeroTangentVectorInitializer(),
  //                   y: y.zeroTangentVectorInitializer())
  //   }
  // }

  // Now synthesized, correct:
  // var zeroTangentVectorInitializer: () -> TangentVector {
  //   { [xZeroTanInit = x.zeroTangentVectorInitializer,
  //      yZeroTanInit = y.zeroTangentVectorInitializer] in
  //     TangentVector(x: xZeroTanInit(), y: yZeroTanInit())
  //   }
  // }
}
```

</details>

Ready for re-review.

dan-zheng commented:
@swift-ci Please smoke test

rxwei commented on test/AutoDiff/Sema/DerivedConformances/derived_zero_tangent_vector_initializer.swift:
```diff
@@ -0,0 +1,106 @@
+
+import _Differentiation
+
+struct MemberwiseTangentVector: Differentiable {
```
This PR seems to be missing `class` tests.

dan-zheng replied:
Added `zeroTangentVectorInitializer` SILGen tests for classes and an enum.

dan-zheng committed:
Add `zeroTangentVectorInitializer` SILGen tests for classes and enums.

dan-zheng committed:
Add `zeroTangentVectorInitializer` SILGen tests for enum.

Add test for enum with custom non-`Self` `TangentVector` type.
Fallback `{ TangentVector.zero }` synthesis is expected and correct.

dan-zheng commented:
@swift-ci Please smoke test

dan-zheng commented:
@swift-ci Please clean smoke test Linux
